File: src/utils/dedupe.ts
import type { Assignment } from '../types/assignments';
import type { Quiz } from '../types/quizzes';
import type { GDB } from '../types/gdb';
import type { Lecture } from '../types/lectures';
import type { UnifiedActivity } from '../types/activities';

export type ActivityKey = string;

export function assignmentKey(a: Assignment): ActivityKey {
  const due = a.dueDate?.toISOString().slice(0, 10) || '';
  return `${a.courseCode}|${a.title}|${due}|${a.totalMarks || ''}`;
}

export function quizKey(q: Quiz): ActivityKey {
  const start = q.startDate?.toISOString().slice(0, 10) || '';
  return `${q.courseCode}|${q.title}|${start}`;
}

export function gdbKey(g: GDB): ActivityKey {
  const due = g.dueDate?.toISOString().slice(0, 10) || '';
  return `${g.courseCode}|${g.title}|${due}`;
}

export function lectureKey(l: Lecture): ActivityKey {
  return `${l.courseCode}|${l.title}|${l.week || ''}|${l.type || ''}`;
}

export function unifiedActivityKey(a: UnifiedActivity): ActivityKey {
  const due = a.dueDate?.toISOString().slice(0, 10) || '';
  return `${a.type}|${a.courseCode}|${a.title}|${due}`;
}

export function dedupe<T>(items: T[], keyFn: (item: T) => ActivityKey): { unique: T[]; duplicates: number } {
  const seen = new Set<ActivityKey>();
  const unique: T[] = [];
  let duplicates = 0;

  for (const item of items) {
    const key = keyFn(item);
    if (seen.has(key)) {
      duplicates++;
    } else {
      seen.add(key);
      unique.push(item);
    }
  }

  return { unique, duplicates };
}

export function dedupeAssignments(assignments: Assignment[]): { unique: Assignment[]; duplicates: number } {
  return dedupe(assignments, assignmentKey);
}

export function dedupeQuizzes(quizzes: Quiz[]): { unique: Quiz[]; duplicates: number } {
  return dedupe(quizzes, quizKey);
}

export function dedupeGDBs(gdbs: GDB[]): { unique: GDB[]; duplicates: number } {
  return dedupe(gdbs, gdbKey);
}

export function dedupeLectures(lectures: Lecture[]): { unique: Lecture[]; duplicates: number } {
  return dedupe(lectures, lectureKey);
}

export function dedupeUnifiedActivities(activities: UnifiedActivity[]): { unique: UnifiedActivity[]; duplicates: number } {
  return dedupe(activities, unifiedActivityKey);
}

File: vulms_dart/lib/src/utils/dedupe.dart
/// Utility functions for deduplicating activity lists.
class Dedupe {
  Dedupe._();

  /// Deduplicate items using a key function.
  static DedupeResult<T> dedupe<T>(List<T> items, String Function(T) keyFn) {
    final seen = <String>{};
    final unique = <T>[];
    var duplicates = 0;

    for (final item in items) {
      final key = keyFn(item);
      if (seen.contains(key)) {
        duplicates++;
      } else {
        seen.add(key);
        unique.add(item);
      }
    }

    return DedupeResult(unique: unique, duplicates: duplicates);
  }

  /// Generate a deduplication key for an assignment.
  static String assignmentKey(String courseCode, String title,
      DateTime? dueDate, double? totalMarks) {
    final due = dueDate?.toIso8601String().substring(0, 10) ?? '';
    return '$courseCode|$title|$due|${totalMarks ?? ''}';
  }

  /// Generate a deduplication key for a quiz.
  static String quizKey(String courseCode, String title, DateTime? startDate) {
    final start = startDate?.toIso8601String().substring(0, 10) ?? '';
    return '$courseCode|$title|$start';
  }

  /// Generate a deduplication key for a GDB.
  static String gdbKey(String courseCode, String title, DateTime? dueDate) {
    final due = dueDate?.toIso8601String().substring(0, 10) ?? '';
    return '$courseCode|$title|$due';
  }

  /// Generate a deduplication key for a lecture.
  static String lectureKey(
      String courseCode, String title, int? week, String? type) {
    return '$courseCode|$title|${week ?? ''}|${type ?? ''}';
  }
}

/// Result of a deduplication operation.
class DedupeResult<T> {
  final List<T> unique;
  final int duplicates;

  const DedupeResult({required this.unique, required this.duplicates});
}
